internal/ingress/controller/controller.go

```diff
@@ -725,6 +744,13 @@ func (n *NGINXController) createUpstreams(data []*extensions.Ingress, du *ingres
 					}
 				}
 
+				// marks the upstream as virtual for association with real upstream
+				if anns.Canary.Enabled {
```

Comment:
Why not:
```
upstreams[name].Virtual = anns.Canary.Enabled
```

Reply:
Now that the virtual backend metadata is set when this check happens, the if statement is needed.